### ISSUE TITLE
Roslyn workspace should send notifications when it has completed the loading of all ....

### ISSUE THREAD

**Issue by ManishJayaswal** [dotnet/roslyn#15612]
projects in Light Weight solution load mode. Else LUT does not know that all projects are loaded and we show uncovered glyphs.

**Comment by CyrusNajmabadi** [comment]
Can you not use: KnownUIContexts.SolutionExistsAndFullyLoadedContext?

**Comment by Pilchie** [comment]
No, this is for lightweight solution load - LUT wants to know when Roslyn's background population of the workspace is finished.

**Comment by CyrusNajmabadi** [comment]
Is Workspace.WorkspaceEvents not enough?

**Comment by ManishJayaswal** [comment]
LUT currently does not listen to workspace events directly. It responds to solution crawler events. I am not sure if solution crawler passes this all the time. If it does then maybe LUT first needs to look for this event before starting in lightweight solution load mode.

@KevinH-MS @JohnHamby thoughts?

**Comment by CyrusNajmabadi** [comment]
> LUT currently does not listen to workspace events directly. 

But this bug is literally titled: 

> Roslyn workspace should send notifications when it has completed the loading

The Roslyn workspace does send events.  Through Workspace.WorkspaceEvents :)

Why not just listen to them?

**Comment by CyrusNajmabadi** [comment]
> It responds to solution crawler events

Are you an IIncrementalAnalyzer?

**Comment by ManishJayaswal** [comment]
Yes we are using IIncrementalAnalyzer.
We had gone the solution crawler route because it is already tuned to not impact typing responsiveness etc. which we needed for LUT as well. If Roslyn workspace does send an event when it is completely populated in lightweight solution load mode and the solution crawler does pass it along then there may not be anything to do here on Roslyn side. Can you confirm that workspace event is sent when it is fully populated in lightweight solution load mode?

**Comment by Pilchie** [comment]
You should confirm, but I think you will get a `SolutionAdded` event when the workspace first gets populated.

**Comment by CyrusNajmabadi** [comment]
>Yes we are using IIncrementalAnalyzer.

"Would listening to NewSolutionSnapshotAsync work?"

Why not just listen to the IIncrementalAnalyzer events, aggregating the messages you're hearing, then wait X number of seconds until the last one comes in, and then start processing them?  You can see an example of how we do that in https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs .  

Specifically: https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs#L275

Here we listen to workspace events and aggregate the changes.  We continually kick of a job to process those changes 1-second in the future.  If more events come in, we postpone processing for another second.  This helps as many events come in flurries.  Once we actually start processing, we then go through the list of work we were building up.

**Comment by ManishJayaswal** [comment]
@drognanar can you take a look at this code to see if you are doing something similar when listening for workspace events? Additionally can you also add the ability in LUT to listen to events that indicate that the roslyn workspace is fully loaded . We should probably not start LUT till we receive this event.
We would still not get full coverage if there are non VB/C# projects in the solution which are not represented in roslyn workspace and which have not been loaded by VS. We should probably detect those cases and force load everything in those cases.

**Comment by Pilchie** [comment]
Closing this for now - let us know if WorkspaceEvents aren't enough
